### PR TITLE
[Fix #5527] Allow object check and method chain calling nil object

### LINKED ISSUE #5527: Style/SafeNavigation does't understand order of operations
--------

## Expected behavior

This  should't be flagged as an error

## Actual behavior

```
-      while ref && ref.children.length > 0
+      while ref&.children&.length > 0
```

Breaks the code with:

```
undefined method `>' for nil:NilClass
```

## RuboCop version

Include the output of `rubocop -V`. Here's an example:

```
$ rubocop -V
0.52.1 (using Parser 2.4.0.2, running on ruby 2.4.1 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by jonas054**
Expressions such as

```ruby
  foo && foo.bar.some_method
```

cannot always be changed into code using the safe navigation operator `&.`. If `some_method` is a method that can be called on `nil` (e.g., `nil?`), then we're out of luck. It's because `nil && nil.bar.nil?` returns `nil`, while `nil&.bar.nil?` returns `true`.

A related problem is that if there's a chain of methods and the `Style/SafeNavigation` cop converts the first `.` into `&.` (or the user does so, following the cop's advice), then we depend on the `Lint/SafeNavigationChain` to convert the rest of the chain. If `Lint/SafeNavigationChain` is disabled we can't risk making a semantic change, so `Lint/SafeNavigationChain` doesn't report an offense in that situation.

### CODE REVIEW

File: spec/rubocop/cop/style/safe_navigation_spec.rb [hunk 1]
```diff
@@ -416,14 +431,39 @@
...
+                                    },
+                                    'Lint/SafeNavigationChain' => {
+                                      'Enabled' => false
+                                    },
+                                    'Style/SafeNavigation' => cop_config)
+              end
+
+              it 'allows an object check followed by chained method calls' do
+                expect_no_offenses(<<-RUBY.strip_indent)
+                  #{variable} && #{variable}.one.two.three(baz) { |e| e.qux }
```
Comment (garettarrowood) [inline spec/rubocop/cop/style/safe_navigation_spec.rb]: Looks like the implied string interpolation is making this just a comment.

**Comment by bbatsov** [comment]
@rrosenblum Please, take a look as well. 

Seems our problems with this cop are never ending. I'm starting to wonder if we'll ever get a handle of all of them.

File: spec/rubocop/cop/style/safe_navigation_spec.rb (hunk 1 above)
Comment (garettarrowood) [reply spec/rubocop/cop/style/safe_navigation_spec.rb]: Nice! Thank you, I was not aware of that behavior with heredocs.

File: lib/rubocop/cop/style/safe_navigation.rb [hunk 2]
```diff
@@ -176,7 +177,12 @@ def unsafe_method_used?(method_chain, method)
           return true if unsafe_method?(method)
 
           method.each_ancestor(:send).any? do |ancestor|
+            unless config.for_cop('Lint/SafeNavigationChain')['Enabled']
```
Comment (rrosenblum) [inline lib/rubocop/cop/style/safe_navigation.rb]: I like this addition.

**Comment by rrosenblum** [comment]
> Seems our problems with this cop are never ending. I'm starting to wonder if we'll ever get a handle of all of them.

I agree. We never expected to have so many issues with this cop. This has been more complicated than originally expected. I am starting to think that we should start restricting the scope of this cop. For guaranteed safety, we could stop registering an offense any time there is a method chain, or only allow correction when there is at most 2 methods. I think there is a chance for some issues with long method chains, `foo && foo.bar.baz.qux` and `foo && foo.bar.baz.nil?`. To maintain the same functionality of this code, we would only be able to use safe navigation on the first call `foo&.bar`. If `bar` happens to be `nil`, this will raise a `NoMethodError` with or without safe navigation. I am beginning to think that we should remove auto-correct from `Lint/SafeNavigationChain` completely and force users to handle this situation themselves. In the chaining scenario just mentioned `Lint/SafeNavigationChain` would auto-correct and change the potential output of the methods. Maybe the concept of `Lint/SafeNavigationChain` should be changed to warning about potential `NoMethodErrors` based on chaining of methods that don't have complete `nil` checks. Allow method chains with no safety. Register an offense for `foo&.bar.baz`, `foo && foo.bar.baz`, `foo.bar.baz if foo`, and other similar code. 

I also want to introduce a new cop to make sure safe navigation is balanced in `&&` and `||` statements. We overlooked some edge cases for consistency with `foo && foo.bar && foo.baz`. This will be corrected to `foo&.bar && foo.baz`.

**Comment by jonas054** [comment]
@rrosenblum Do you have time to implement these ideas yourself? Feel free to use anything you like from my PR. I'll close it when you submit yours.

**Comment by bbatsov** [comment]
I plan to cut a new release in the next couple of days, so I'll merge it for now.

**Comment by rrosenblum** [comment]
@jonas054 I'll get started on the changes mentioned above.

**Pull request merged**